### PR TITLE
Cross-platform targeting feedback

### PR DESCRIPTION
**Pull request by JamesNK**
@terrajobst Your cross-platform targeting feedback (and some other bits) applied

TODO:
- [ ] A section that talks about to wrap platfom-specific code via a .NET Standard library
- [x] We should call out that strong naming on .NET Framework is viral, which is why customers often want libraries to be strong named so that they can be strong named and be able to GAC their assemblies.
- [x] Separately, we should emphasize that this is .NET Framework only problem and that libraries that aren't designed to ever work on .NET Framework don't have to be signed (for example, all libraries offering mobile functionality).
- [x] We should generally replace all usages of modern .NET Class Library with SDK-style projects. **What should these project files be called?**
- [ ] We should clarify that these guidelines mostly apply to OSS on NuGet.org and closed source with private package feeds (Azure Artifacts or MyGet).
- [x] We should call out that there should be a relationship between the NuGet version and the assembly version (i.e. aligning at least the major version, but ideally major & minor).
- [x] We should explicitly call that customers should not have a fixed file version as it might break setups.
- [ ] We might want to include pointers to other tools, such as Nerdbank.GitVersioning for generating version numbers.
- [ ] We should try to explain why breaking changes are bad, in particular that there might be certain package graphs where you cannot select a version of the package that works for everyone.

### CODE REVIEW

**Comment by JamesNK** [comment]
> CONSIDER using SemVer 2.0.0 to version your NuGet package.
> We should make sure that this doesn't cut off large customer segments due to the higher NuGet client. Have we reviewed this with the NuGet folks? They likely have that information.

They have already reviewed it and didn't call this out a problem.

**Comment by JamesNK** [comment]
> A section that talks about to wrap platfom-specific code via a .NET Standard library

Everything on the current page focuses on choosing targets. Implementation guidance is different enough that I think it could be a new page, and it prevents the current one from being too long

@terrajobst if you could put the new page together that would be great.

**Comment by JamesNK** [comment]
Other items have been moved to issues.

File: docs/cross-platform-targeting.md [hunk 1]
```diff
@@ -64,6 +35,48 @@ Targeting .NET Standard, and successfully compiling your project, does not guara
 * [.NET Standard](https://docs.microsoft.com/en-us/dotnet/standard/net-standard)
 * [.NET API analyzer](https://docs.microsoft.com/en-us/dotnet/standard/analyzers/api-analyzer)
 
+## Multi-targeting
+
+An alternative way to add cross-platform support to a .NET library is multi-targeting. Multi-targeting involves targeting .NET implementations individually and includes multiple assemblies in a NuGet package. This leverages NuGet's ability to have multiple assemblies in a package and select the best one when added to an application.
+
+You can combine multi-targeting with .NET Standard. For example, a `netstandard2.0` target could be the default implementation and you could provide a .NET implementation specific target that uses implementation specific APIs for additional features. NuGet will automatically select the implementation specific target when possible, e.g. a .NET Core application will use the `netcoreapp2.0` assembly over the `netstandard2.0` assembly in a NuGet package.
+
```
Comment (terrajobst) [inline docs/cross-platform-targeting.md]: I don't think we should position multi-targeting as an *alternative* to .NET Standard. While it *could* be used this way, that's not what we recommend because it forwards the multi-targeting requirement to the consumer. Rather, our recommendation is to build bait-and-switch-style .NET Standard libraries where you have always target .NET Standard while *also* providing framework-specific implementations.

Hence, I suggest we reword this to something like this:

> Sometimes you need to access framework-specific APIs from your libraries. The best way to do this is by using *multi-targeting*, which builds your project  for multiple frameworks rather than for just one.
>
> In order to shield your consumers from having to build for multiple frameworks too you should always produce a .NET Standard output plus one or more framework-specific outputs. All outputs will be packaged inside a single NuGet package, allowing consumers to reference the same package and let NuGet pick the appropriate implementation. Your .NET Standard library serves as the fall-back library that is used everywhere, except for the cases where your NuGet package offers a framework-specific implementation. This enables you to use conditional compilation in your code in order to call framework-specific APIs.

File: docs/cross-platform-targeting.md [hunk 2]
```diff
@@ -64,6 +35,48 @@ Targeting .NET Standard, and successfully compiling your project, does not guara
...
+**✔️ DO** use a project file's `TargetFrameworks` property when multi-targeting.
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <!-- This project will output netstandard2.0 and net471 assemblies -->
+    <TargetFrameworks>netstandard2.0;net471</TargetFrameworks>
+  </PropertyGroup>
+</Project>
+```
```
Comment (terrajobst) [inline docs/cross-platform-targeting.md]: We should add a plug for Oren's package here:

* **✔️ CONSIDER** using [MSBuild.Sdk.Extras](https://github.com/onovotny/MSBuildSdkExtras) when multi-targeting for UWP and Xamarin as it greatly simplifies your project file.

File: docs/cross-platform-targeting.md [hunk 3]
```diff
@@ -64,6 +35,48 @@ Targeting .NET Standard, and successfully compiling your project, does not guara
...
+
+**More Information**
+
+* [.NET target frameworks](https://docs.microsoft.com/en-us/dotnet/standard/frameworks)
+* [Multi-Targeting and Porting a .NET Library to .NET Core 2.0](https://weblog.west-wind.com/posts/2017/Jun/22/MultiTargeting-and-Porting-a-NET-Library-to-NET-Core-20)
+
 ## Older Targets
 
 .NET supports targeting versions of the .NET Framework that are long out of support, e.g. .NET 2.0, as well as platforms that are no longer commonly used, e.g. Silverlight and Windows Phone. The value of targeting such old platforms can be easily outweighed by the overhead of programming around missing APIs.
```
Comment (terrajobst) [inline docs/cross-platform-targeting.md]: I found this sentence hard to parse:

>The value of targeting such old platforms can be easily outweighed by the overhead of programming around missing APIs.

How about:

>While is there is value in making your library work on as many targets as possible, there is also significant overhead if you have to workaround missing APIs. We believe certain frameworks are no longer worth targeting, considering their reach and limitations:

**Comment by terrajobst** [comment]
Just saw I didn't submit my review this morning. I'll submit PRs for the suggestions.

**Comment by terrajobst** [comment]
Submitted #27

**Comment by JamesNK** [comment]
All good points. I merged because I submitted the dotnet/docs PR. That will be the best place for future changes to avoid these two places getting out of sync.
